kube-proxy registers its version

fraenkel opened the pull request:
Rather than have a kubelet register its version as the kube-proxy
version, have the kube-proxy do it as part of its start up.

fixes #48811

xiangpengzhao commented on cmd/kube-proxy/app/server.go:
```diff
@@ -774,6 +779,50 @@ func tryIPTablesProxy(iptver iptables.IPTablesVersioner, kcompat iptables.Kernel
+	go wait.Until(func() {
+		oldNode, err := s.CoreClient.Nodes().Get(nodeName, metav1.GetOptions{})
+		if err != nil {
+			glog.Warningf("Failed to retrieve node info: %v", err)
```
We may get spam logs in some case. See the race between kube-proxy start up and node registration #37414.

So I suggest the way #47262 did.

fraenkel committed:
kube-proxy registers its version

Rather than have a kubelet register its version as the kube-proxy
version, have the kube-proxy do it as part of its start up.

ncdc commented on cmd/kube-proxy/app/server.go:
```diff
@@ -383,7 +386,7 @@ type ProxyServer struct {
 // createClients creates a kube client and an event client from the given config and masterOverride.
 // TODO remove masterOverride when CLI flags are removed.
-func createClients(config componentconfig.ClientConnectionConfiguration, masterOverride string) (clientset.Interface, v1core.EventsGetter, error) {
+func createClients(config componentconfig.ClientConnectionConfiguration, masterOverride string) (clientset.Interface, v1core.CoreV1Interface, error) {
```
Why not just use the first return value - `clientset.Interface` - for reporting the version?

fraenkel replied:
I tried using api.Node instead of v1.Node when creating the TwoWayMerge which ended up with a 200 and nothing changed. The api.Node doesn't have any of the json annotation so the patch is incorrect. I am open to other suggestions.

ncdc replied:
Sorry, I missed that `clientset` is internal.

ncdc commented on cmd/kube-proxy/app/server.go:
```diff
@@ -792,3 +797,58 @@ func getNodeIP(client clientset.Interface, hostname string) net.IP {
+func reportVersion(client v1core.CoreV1Interface, nodeName string) error {
+	v := version.Get().String()
+	nodeClient := client.Nodes()
+	backoff := wait.Backoff{
```
Could you use `retry.DefaultBackoff` instead?

fraenkel replied:
Sure, was just going with the original comment which had me copy other code. I wasn't even sure if we care if this succeeds. Right now its blocking and ignores the error. We could either have it attempt this in the background and ignore the error or attempt and die if it never succeeds.

ncdc commented on cmd/kube-proxy/app/server.go:
```diff
@@ -792,3 +797,58 @@ func getNodeIP(client clientset.Interface, hostname string) net.IP {
+		}
+		return true, nil
+	}
+	err := wait.ExponentialBackoff(backoff, tryUpdateStatus)
```
I'd recommend using `retry.RetryOnConflict`

ncdc reviewed (changes requested):
Could you please also consider adding an integration test that verifies that kube-proxy reports its version and updates the node's status when it runs?

fraenkel commented:
@ncdc Struggling with an integration test. I have it written but alas I haven't figured out a good way to actually run it. The proxier requires ROOT if I run in IPTables mode. If I run in userspace mode, I blow up on setting the RLimit.
Open to suggestions on how to test this.

ncdc commented:
@fraenkel ah I had forgotten about the root requirement. You could make it an e2e, or you could construct a `*ProxyServer` manually instead of calling `NewProxyServer`. As long as you fill in the fields such that you avoid nil pointer exceptions, you can fill them with fake/mock implementations of the various interfaces (such as `Proxier`), and then I think you can get by without needing root.

fraenkel commented:
@ncdc If I build my own ProxyServer, this isn't really an integration test. This also requires the version reporting to occur during Run() which I have no problem with, it was there then I moved it but I can put it back.
We still need to decide what we do if it fails. Right now its synchronous which doesn't make sense if we don't care about it failing. If the Run() should fail if the version cannot be reported then I just need to add that check and leave it synchronous.
I am leaning toward e2e because that will test that the version is set. What it doesn't prove is that it was empty prior to the test.

ncdc commented:
Ok, how about just a unit test?

fraenkel commented:
That was the next choice.
Do you have a preference on the behavior of success/failure? And whether its done during construction vs Run()?

ncdc commented:
I'm not sure what the implications might be (if any) if kube-proxy fails to report its version. Do you know who uses this information?

We want to avoid the constructor from having side effects, so the version should be reported in Run().

fraenkel commented:
@ncdc Unit was added, and e2e is failing.
gce-etcd3 fails due to RBAC, can't update node/status. I can track this one down.
For kops, it looks like the node name isn't really determined. Guess this was always the case but no one cared until now.

ncdc commented on cmd/kube-proxy/app/server.go:
```diff
@@ -778,6 +784,54 @@ func (s *ProxyServer) birthCry() {
+			glog.Warningf("Failed to retrieve node info: %v", err)
+			return err
+		}
+		clonedNode, err := api.Scheme.DeepCopy(oldNode)
```
Do you need to do a deep copy, since you're using a client to retrieve oldNode? Could you get oldNode, marshal to oldData, set the version, marshal to newData, and go from there?

fraenkel replied:
This I copied from elsewhere. I will optimize it for this usecase.

ncdc commented on cmd/kube-proxy/app/server_test.go:
```diff
@@ -396,3 +415,36 @@ func TestLoadConfigFailures(t *testing.T) {
+	server.Run()
+	select {
+	case <-nodePatched:
+	case <-time.After(time.Second):
```
time.Second could end up being too small. Please use wait.ForeverTestTimeout instead.

fraenkel replied:
this is all internal so it should be immediate but I will change it.

ncdc replied:
Trust me, you'll be happier using a longer timeout to avoid potential flakes due to cpu contention :-)

ncdc commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go:
```diff
@@ -268,7 +268,7 @@ func ClusterRoles() []rbac.ClusterRole {
 				rbac.NewRule("list", "watch").Groups(legacyGroup).Resources("services", "endpoints").RuleOrDie(),
 				rbac.NewRule("get").Groups(legacyGroup).Resources("nodes").RuleOrDie(),
-
+				rbac.NewRule("patch").Groups(legacyGroup).Resources("nodes/status").RuleOrDie(),
```
FYI @kubernetes/sig-auth-misc @liggitt @deads2k @jbeda @smarterclayton re allowing kube-proxy to patch nodes/status.

liggitt replied:
That's unexpected...

deads2k replied:
> FYI @kubernetes/sig-auth-misc @liggitt @deads2k @jbeda @smarterclayton re allowing kube-proxy to patch nodes/status.

Patching node status allows you to change your labels, which allows you to steer pod scheduling, which probably isn't something to give `kube-proxy`.

fraenkel replied:
I am open to suggestions. The KubeProxyVersion lives in an awkward place that requires an update. My only other idea was to do the reverse and have the kubelet contact the kube-proxy which I thought was a bit worse.

jbeda commented:
I was talking to someone the other day (don't remember who) and they had the insight that the Node object should really be named the Kubelet object.  It seems really wacky to have the proxy updating the node.

Some observations:
* Many people (including the kubeadm flow) launch the proxy as a DaemonSet.  The version is reported and controlled through that.  In this way the proxy is no different than any other DaemonSet.
* Anyone running a production solution should have monitoring in place.  Having versions exported as metrics would help there.
* There is no reason that the proxy can't run on machines that don't have a kubelet running on them.  You may want to have machines that have ClusterIPs configured but don't take on workloads.
* The strategic merge patch flow is very confusing and non-obvious.  Relying on it for something like this, at least to me, smells really bad.  Having two components editing fine grained parts of an object should be avoided.

I'd suggest we go in a different direction and remove/deprecate the proxy version from NodeSystemInfo.  Or, failing that, have the kubelet otherwise communicate locally with the KubeProxy (look for a file dropped in a well known place) to find version information.

yujuhong commented:
/cc @kubernetes/sig-network-misc since this changes kube-proxy.

> I'd suggest we go in a different direction and remove/deprecate the proxy version from NodeSystemInfo. Or, failing that, have the kubelet otherwise communicate locally with the KubeProxy (look for a file dropped in a well known place?) to find version information.

+1 for deprecating this field. 
FWIW, sig-network is going to move kube-proxy into a DaemonSet soon (#23225). I don't think it makes sense for kubelet to continue reporting the version for kube-proxy, when it may not even be running on the node.

Maybe @dchen1107 would like to weigh in.
